src/main/java/com/lebediev/movieland/service/conversion/CachedCurrency.java

```diff
@@ -0,0 +1,93 @@
+import java.util.ArrayList;
+import java.util.List;
+
+@Service
+public class CachedCurrency {
+    private final static Logger log = LoggerFactory.getLogger(CachedCurrency.class);
+    private volatile List<CurrencyEntity> currencyList;
+    private final static DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd.MM.yyyy");
+    @Value("${currency.url}")
+    private String url; //= "https://bank.gov.ua/NBUStatService/v1/statdirectory/exchange?json";
+
+    public CurrencyEntity getCurrencyEntity(Currency currency){
+        log.info("Start getting currency from cache");
+        long startTime = System.currentTimeMillis();
+        List<CurrencyEntity> currencyEntityListCached = new ArrayList<>(currencyList);
+        for(CurrencyEntity currencyEntity : currencyEntityListCached){
+            if(currencyEntity.getCurrency() == currency){
+                log.info("Finish getting currency from cache. It took {} ms", System.currentTimeMillis() - startTime);
+                return currencyEntity;
+            }
+        }
+        throw new RuntimeException("Could not get cached currency " + currency);
+    }
+
+
+    public static List<CurrencyEntity> getCurrenciesFromUrl(String url){
+        log.info("Start getting currency list from url = {}", url);
+        long startTime = System.currentTimeMillis();
+        ObjectMapper objectMapper = new ObjectMapper();
+        List<CurrencyEntity> currencyEntityList = new ArrayList<>();
+
+        try{
+            Security.insertProviderAt(new BouncyCastleProvider(), 1);
+
+            URL urlReady = new URL(url);
+            URLConnection connection = urlReady.openConnection();
+            BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
+
+            JsonNode rootNode = objectMapper.readTree(reader);
+            for (JsonNode node : rootNode) {
+                String currencyName = node.get("cc").asText();
+                if (Currency.isValid(currencyName)) {
+                    LocalDate date = LocalDate.parse(node.get("exchangedate").asText(), formatter);
+                    CurrencyEntity currencyEntity = new CurrencyEntity(Currency.getCurrency(currencyName), node.get("rate").asDouble(), date);
+                    currencyEntityList.add(currencyEntity);
+                }
+            }
+            if(currencyEntityList.size() == 0){
+                log.error("Could not get currency list from url: " + url);
+                throw new RuntimeException("Nothing to return - list is empty");
+            }
+            reader.close();
+            log.info("Finish getting currency list from url: {}. It took {} ms",url, System.currentTimeMillis() - startTime);
+            return currencyEntityList;
+
+        }  catch (IOException e) {
+            throw new RuntimeException("Error in connection for url = " + url + " : " + e);
+        }
+    }
+
+    public static boolean isExpired(CurrencyEntity currencyEntity){
```

Comment:
url returns CCYs for current day, do you really want to check it for expiration?

Reply:
for case when we could not obtain actual data and our cache will contain expired rates